### PR TITLE
Implement `prove_substitutions` in `SimplificationPerformer`

### PR DESCRIPTION
**Pull request by fiedlr**

### CODE REVIEW

File: generation/src/proof_generation/k/execution_proof_generation.py [hunk 1]
```diff
@@ -71,12 +71,8 @@ def equality_proof(
     def equality_transitivity(self, last_proof: proof.ProofThunk, new_proof: proof.ProofThunk) -> proof.ProofThunk:
         return self.kore_lemmas.sorted_eq_trans(last_proof, new_proof)
 
-    def prove_substitutions(self, pattern: proof.ProofThunk, substitutions: dict[int, Pattern]) -> proof.ProofThunk:
-        # TODO: Get functional axioms (check out the proof expr below)
-        # TODO: Apply the substitution lemma
-        # TODO: Using the hacky temporary implementation until we don't have the substitutions
-        substituted = pattern.conc.apply_esubsts(substitutions)
-        return proof.ProofThunk((lambda interpreter: proof.Proved(substituted)), substituted)
+    def prove_substitutions(self, pattern_pt: proof.ProofThunk, substitutions: dict[int, Pattern]) -> proof.ProofThunk:
```
Comment (IljaZakharov) [inline generation/src/proof_generation/k/execution_proof_generation.py:74]: Most of proof thunks are named as something plus the suffix `proof` or just `proof`.  The current name looks a bit misleading to me, as it is not just a pattern.

File: generation/src/proof_generation/k/execution_proof_generation.py [hunk 2]
```diff
@@ -385,6 +381,30 @@ def assert_functional_pattern(semantics: LanguageSemantics, pattern: Pattern) ->
         assert k_sym is not None
         assert k_sym.is_functional
 
+    @staticmethod
+    def prove_substitutions(
+        prover: ExecutionProofExp | SimplificationProver, pattern_pf: proof.ProofThunk, substitution: dict[int, Pattern]
```
Comment (IljaZakharov) [inline generation/src/proof_generation/k/execution_proof_generation.py:386]: Same about `pattern_pt`.

File: generation/src/proof_generation/k/execution_proof_generation.py [hunk 3]
```diff
@@ -385,6 +381,30 @@ def assert_functional_pattern(semantics: LanguageSemantics, pattern: Pattern) ->
...
+        """
+        for name, plug in substitution.items():
+            ExecutionProofExp.assert_functional_pattern(prover.language_semantics, plug)
+            prover.add_assumption(functional(0)(plug))
+            # TODO: Get rid of this hack:
+            prover.subst_proofexp.add_axiom(func_subst_axiom(name))
+
+        # Compute the proof
+        step_pf = pattern_pf
+        for name, plug in substitution.items():
```
Comment (IljaZakharov) [inline generation/src/proof_generation/k/execution_proof_generation.py:401]: We need to iterate over substitutions in particular order (increasing). If a rule is already universally generalized, we expect a particular outer `forall` quantifier.

Comment (IljaZakharov) [reply generation/src/proof_generation/k/execution_proof_generation.py:401]: But I can add this sorting in my PR related to making these rules generalized later.

File: generation/src/proof_generation/k/execution_proof_generation.py [hunk 4]
```diff
@@ -399,29 +419,20 @@ def rewrite_event(self, rule: KRewritingRule, substitution: dict[int, Pattern])
...
-            self.subst_proofexp.add_axiom(func_subst_axiom(name))
 
-        # Compute the proof
-        step_pf = self.load_axiom(rule.pattern)
-        for name, plug in substitution.items():
-            functional_assumption = self.load_axiom(functional(0)(plug))
-            universalized = self.subst_proofexp.universal_gen(step_pf, EVar(name))
-            step_pf = self.subst_proofexp.functional_subst(functional_assumption, universalized, EVar(name))
+        # Prove the instance of the used rule axiom
+        pf = ExecutionProofExp.prove_substitutions(self, self.load_axiom(rule.pattern), substitution)
```
Comment (IljaZakharov) [inline generation/src/proof_generation/k/execution_proof_generation.py:426]: Let's call it `proof`?

**Review by IljaZakharov — APPROVED** [review]
Looks nice! I left a couple of comments though.

File: generation/src/proof_generation/k/execution_proof_generation.py (hunk 3 above)
Comment (fiedlr) [reply generation/src/proof_generation/k/execution_proof_generation.py:401]: That's true but this is meant to be merged before your PR

**Pull request merged**